### ISSUE TITLE
Implement http.ResponseWriter to capture status code

### ISSUE THREAD

**Issue by cbandy** [appneta/go-traceview#1]
We can intercept the response status code with a proxy `http.ResponseWriter`. Something like this:

```diff
diff --git a/traceview/http_instrumentation.go b/traceview/http_instrumentation.go
index 8a09afb..8374c7d 100644
--- a/traceview/http_instrumentation.go
+++ b/traceview/http_instrumentation.go
@@ -38,15 +38,16 @@ func InstrumentedHttpHandler(handler func(http.ResponseWriter, *http.Request)) f
 
 			// Create exit event, but do not report yet. Just add its X-Trace header:
 			event = exitEvent(ctx)
-			w.Header()["X-Trace"] = []string{event.MetaDataString()}
+			w.Header().Set("X-Trace", event.MetaDataString())
+
+			w = httpResponseWriter{w, event}
 		}
 
 		// Call original HTTP handler:
 		handler(w, r)
 
 		if sampled {
-			// Add status code and report exit event:
-			event.AddInt64("Status", getStatusCode(w))
+			// Report exit event:
 			event.Report(ctx)
 		}
 
@@ -78,21 +79,12 @@ func exitEvent(ctx *Context) *Event {
 	return e
 }
 
-// Extracts HTTP status code from writer, using reflection. This depends on http.response internals.
-// Returns -1 if status was not found.
-func getStatusCode(w http.ResponseWriter) int64 {
-	var status int64 = -1
-
-	ptr := reflect.ValueOf(w)
-	if ptr.Kind() == reflect.Ptr {
-		val := ptr.Elem()
-		if val.Kind() == reflect.Struct {
-			field := val.FieldByName("status")
-			if field.Kind() == reflect.Int {
-				status = field.Int()
-			}
-		}
-	}
+type httpResponseWriter struct {
+	http.ResponseWriter
+	event Event
+}
 
-	return status
+func (w httpResponseWriter) WriteHeader(status int) {
+	w.ResponseWriter.WriteHeader(status)
+	w.event.AddInt64("Status", status)
 }
```

I haven't created an account, so I wasn't able to test this.

**Comment by daver76** [comment]
@cbandy Thanks! We'll give it a try and let you know how this approach worked out.

**Comment by cce** [comment]
Fixed by #3.
